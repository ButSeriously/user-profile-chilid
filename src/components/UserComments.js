import React from 'react';
import uuid from 'uuid';
import data from '../../storage/data.json';
import UserComment from './UserComment';

class UserComments extends React.Component {
  state = {
    comments: data.comments,
    user: data.logged_in,
    commentsAmount: data.comments.length,
    areCommentsVisible: true,
    hidingButtonTextContent: 'Hide comments',
  };

  addComment = (e) => {
    e.preventDefault();
    const content = e.target.elements.comment.value.trim();
    e.target.elements.comment.value = ''; // Reset text input value
    if (content) {
      this.setState(prevState => ({
        comments: prevState.comments.concat([{
          photo: this.state.user.photo,
          name: this.state.user.user,
          content,
          date: new Date(),
        }]),
        commentsAmount: prevState.commentsAmount + 1,
        areCommentsVisible: true,
      }));
    }
  }

  hidingButton = () => {
    this.setState(prevState => ({
      areCommentsVisible: !prevState.areCommentsVisible,
      hidingButtonTextContent: this.state.areCommentsVisible ?
        'Show comments' : 'Hide comments',
    }));
  }

  sortCommentsByDate = () =>
    this.state.comments.sort((a, b) =>
      new Date(a.date) - new Date(b.date));

  handleCommentsDisplaying = () => {
    const classList = ['user_comments__scroll'];
    if (!this.state.areCommentsVisible) {
      classList.push('user_comments__scroll--display');
    }
    return classList;
  }

  render() {
    this.sortCommentsByDate();
    return (
      <div className="user_comments">
        <button onClick={this.hidingButton} className="user_comments__hiding">{this.state.hidingButtonTextContent} (
          { this.state.commentsAmount })
        </button>
        <div className={this.handleCommentsDisplaying().join(' ')} >
          { [this.state.comments.map(comment =>
            <UserComment key={uuid()} comment={comment} />)] }
        </div>
        <form onSubmit={this.addComment} className="user_comments__form">
          <input className="user_comments__textbox" type="text" name="comment" placeholder="Add a comment" />
        </form>
      </div>
    );
  }
}

export default UserComments;
