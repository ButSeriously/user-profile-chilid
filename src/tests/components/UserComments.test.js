import React from 'react';
import { shallow } from 'enzyme';
import UserComments from '../../components/UserComments';

test('should render UserComments', () => {
  const wrapper = shallow(<UserComments />);
  expect(wrapper).toMatchSnapshot();
});

test('should handle comment adding', () => {
  const wrapper = shallow(<UserComments />);
  wrapper.setState({ comments: [] });
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
    target: {
      elements: {
        comment: { value: 'My commentary' },
      },
    },
  });
  const newComment = wrapper.state('comments');
  expect(newComment[0].content).toBe('My commentary');
});

test('should not add comment when no value provided', () => {
  const wrapper = shallow(<UserComments />);
  wrapper.setState({ comments: [] });
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
    target: {
      elements: {
        comment: { value: '' },
      },
    },
  });
  expect(wrapper.state('comments')).toHaveLength(0);
});
