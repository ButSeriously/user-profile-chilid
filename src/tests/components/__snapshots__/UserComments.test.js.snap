// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserComments 1`] = `
<div
  className="user_comments"
>
  <button
    className="user_comments__hiding"
    onClick={[Function]}
  >
    Hide comments
     (
    2
    )
  </button>
  <div
    className="user_comments__scroll"
  >
    <UserComment
      comment={
        Object {
          "content": "Lorem ipsum dolor sit amet enim. Etiam ullamcorper. Suspendisse a pellentesque dui, non felis. Maecenas malesuada elit lectus felis, malesuada ultricies. Curabitur et ligula.",
          "date": "2017-11-26T12:45:03.564Z",
          "name": "Mike Ross",
          "photo": "./img/user-photo.jpg",
        }
      }
      key="uuid"
    />
    <UserComment
      comment={
        Object {
          "content": "Lorem ipsum dolor sit amet enim. Etiam ullamcorper. Suspendisse a pellentesque dui, non felis. Maecenas malesuada elit lectus felis, malesuada ultricies. Curabitur et ligula.",
          "date": "2017-11-27T07:11:00.611Z",
          "name": "Mike Ross",
          "photo": "./img/user-photo.jpg",
        }
      }
      key="uuid"
    />
  </div>
  <form
    className="user_comments__form"
    onSubmit={[Function]}
  >
    <input
      className="user_comments__textbox"
      name="comment"
      placeholder="Add a comment"
      type="text"
    />
  </form>
</div>
`;
